eip-0031.md

```diff
@@ -0,0 +1,340 @@
+
+Strong advantage of this approach (compared to Cardano’s one) is that user always knows in advance whether there is an opportunity to pay via “babel fees” or not, and if there is, what is the exchange ratio. He can therefore be (almost) certain that if he decides to use it, his transaction will be included in the blockchain ledger. Be aware, however, that there exist some exceptions to this rule, which is later discussed in the “Wallets implementation” section.
+
+Motivation
+--------------------------------------------
+
+Many users use blockchain solely for transferring (native)tokens, such as stablecoins, or even “meme” coins. These users, understandably, do not want to be bothered with keeping an eye on the amount of blockchain’s primary token they own, or even obtaining this primary token in the first place.
+
+Once they run out of primary token, they have to difficultly (and often costly), swap their tokens of interest for the primary tokens, that they can later use to cover transaction fees. Since primary tokens are also needed for these swaps, users may be forced to introduce new capital to their portfolio solely for the purpose of purchasing primary tokens, used for fee paying. 
+
+Since basic transactional fees on the Ergo blockchain are generally quite low, the "babel fees" users would be probaby willing to pay a fee that could be higher than that of a primary token transaction, in exchange for being able to pay in their token of interest and not having to bother with the blockchain’s primary token purchase.
+
+This brings up a financial incentive for “EIP-0031 supporters”, who could benefit out of this arbitrage by providing the liquidity for such “babel fees” users, with primary token’s selling price (expressed in tokens of interest) being higher compared to the same token pair on the exchanges.
+
+Smartcontract specification
+--------------------------------------------
+Here is the smartcontract's source code in ErgoScript which will be used to protect the babel fee box:
+```scala
+{
+
+    val babelFeeBoxCreator = SELF.R4[SigmaProp].get
+
+    val recreatedBox = OUTPUTS(OUTPUTS.size - 2)
```

Comment:
Why this index chosen?

Reply:
If the tx has a change box for whatever reason then it will be at index OUTPUTS.size-2

Reply:
Edit: I am wrong, the change box is last. Even then though, the miner fee box would end up in index OUTPUTS.size-2

Reply:
Change box is last, that is correct. But change box is not mandatory

Reply:
So if the tx has a change box then you can't use this contract with Appkit, shouldn't EIP contracts take into account use with AppKit, isn't that the point of having a library?

Reply:
With sigma-rust we also haven't a way to set a outputs order, so I just reorder the outputs after transaction is generated.

Reply:
So I have been doing some more work on GuapSwap, and may need to use more than one babel box, depending on the token, in one transaction. So `OUTPUTS(OUTPUTS.size-2)` is very restrictive in this case. Is doing `OUTPUTS.filter{ output: Box => output.propositionBytes == SELF.propositionBytes }(0)` possible instead?

Reply:
Another option is to set output index in context extension of spending input.

Reply:
that's cool way to do it